kong/plugins/tcp-log/handler.lua

```diff
@@ -7,6 +7,27 @@ local ngx = ngx
 local timer_at = ngx.timer.at
 
 
+local function get_body_data()
+  local req = ngx.req
+
+  req.read_body()
+  local data = req.get_body_data()
```

Comment:
any reason for not using pdk functions? https://docs.konghq.com/2.0.x/pdk/kong.request/#kongrequestget_raw_body

Reply:
Good advice. Using PDK functions instead of making my own wheels. The reason I wrote it myself is that I saw the recommendation [here](https://github.com/Kong/kong/pull/3028#discussion_r152298066). Maybe the `kong.request.get_raw_body()` of the pdk did not appear at that time. I will fix it.

Reply:
I did some experiments before. For **grpc requests**, no matter how the `client_body_buffer_size` is set or how small the request body is, the request body will always be stored in a temporary file. This situation has been encountered in both source installation and run docker kong.

This is not a problem of kong, but a bug of nginx processing HTTP/2 request. I saw nginx fixed it: https://nginx.googlesource.com/nginx/+/2939d5c5fadb3f28a8f8ac85c039bfd007239d3c%5E%21/#F0.

But in openresty-1.15.8.2, the nginx code is still the problematic code before the bug fix.

![openresty_nginx](https://user-images.githubusercontent.com/13045944/83780284-8e9f7280-a6bf-11ea-85da-f58376e34fd4.png)

In this case, if you do not read the temporary file, the grpc request body will not be collected. I will continue to dig this problem tomorrow, and if you know more, please let me know ：）

Reply:
This question is related to the following question. Let's discuss how to do better.

Reply:
Fixed. Considering that the grpc request body is not easy to read,  let the backend service log it at this stage. Now use the pdk function `kong.request.get_raw_body()`, which will support logging http1.x request body.